monty-31/src/aarch64_neon/poseidon2.rs

```diff
@@ -318,13 +350,78 @@ where
         external_terminal_permute_state(
             state,
-            self.external_constants.get_terminal_constants(),
-            add_rc_and_sbox_generic,
+            &self.packed_terminal_external_constants,
+            add_rc_and_sbox::<FP, D>,
             &MDSMat4,
         );
     }
 }
 
+/// Converts a scalar constant into a packed NEON vector in "negative form".
+///
+/// This is a pre-computation step for round constants. Instead of storing a constant `c`
+/// and performing a modular addition `(val + c) mod P` in the permutation loop, we pre-compute
+/// `c' = c - P`. This allows the round update to be a simple, non-modular signed integer
+/// addition `val + c'`, which is significantly faster.
+#[inline(always)]
+fn convert_to_vec_neg_form_neon<MP: MontyParameters>(input: i32) -> uint32x4_t {
+    unsafe {
+        // Perform the subtraction as a standard integer operation.
+        //
+        // This computes the "negative form" of the constant.
+        let input_sub_p = input - (MP::PRIME as i32);
+
+        // Broadcast (duplicate) the scalar result into all four lanes of a 128-bit NEON vector.
+        let vec_s32 = aarch64::vdupq_n_s32(input_sub_p);
+
+        // Reinterpret the bits of the signed vector as an unsigned vector.
+        //
+        // This is a zero-cost operation that only changes the type for the Rust compiler.
+        aarch64::vreinterpretq_u32_s32(vec_s32)
+    }
+}
+
+/// Performs the fused AddRoundConstant and S-Box operation `x -> (x + c)^D`.
+///
+/// It uses a pre-computed constant `rc` in a special negative form (`c - P`) for a fast addition.
+/// The addition `val + rc` results in a value in `[-P, P)`, which is passed directly to the
+/// exponentiation function without an intermediate reduction, improving performance.
+///
+/// # Safety
+/// - `val` must contain elements in canonical form `[0, P)`.
+/// - `rc` must contain elements represented as `c - P`, which are in `[-P, 0]`.
+fn add_rc_and_sbox<PMP, const D: u64>(val: &mut PackedMontyField31Neon<PMP>, rc: uint32x4_t)
+where
+    PMP: PackedMontyParameters + FieldParameters,
+{
+    unsafe {
+        // Convert the field element vector to its raw signed i32 representation.
+        let vec_val_s = val.to_signed_vector();
+
+        // Reinterpret the round constant vector as signed. This is safe as `rc = c - P`.
+        let rc_s = aarch64::vreinterpretq_s32_u32(rc);
+
+        // Add the pre-computed `rc = c - P`.
+        //
+        // The result is:
+        // - congruent to `val + c`,
+        // - guaranteed to be in the range `[-P, P)`.
+        let val_plus_rc = aarch64::vaddq_s32(vec_val_s, rc_s);
+
+        // Apply the power S-box `x -> x^D`.
+        //
+        // The `exp_small` function:
+        // - accepts inputs in `(-P, P)`,
+        // - returns a canonical result in `[0, P)`.
+        //
+        // This avoids a costly reduction step.
```

Comment:
We need to update `exp_small` and the similar functions to make it clear that if an input is equal to `-P` that is fine.
```suggestion
        // Apply the power S-box `x -> x^D`.
        //
        // The `exp_small` function:
        // - accepts inputs in `(-P, P)`,
        // - returns a canonical result in `[0, P)`.
```